File: week_8_and_9/3_Ruby/reverse_words/my_solution.rb
# U3.W8-9: Reverse Words


# I worked on this challenge [with: Hing Huynh ].

# 2. Pseudocode
# Split the string input into an array at each space found.
# Use the string reverse method to reverse each string.
# Join the resulting array back into a single string with spaces.
# Return the string.


# 3. Initial Solution

def reverse_words(string)
  array_of_strings = string.split(' ').to_a
  array_of_strings.map! { |string| string.reverse }
  array_of_strings.join(' ')
end

# 4. Refactored Solution

def reverse_words(string)
    string.split(' ').to_a.map! { |string| string.reverse }.join(' ')
end

# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

def assert(statement = "Assertion failed!")
  raise statement unless yield
end

assert { reverse_words("Ich bin ein Berliner") == "hcI nib nie renilreB" }
assert { reverse_words("") == "" }
assert { reverse_words("word") == "word".reverse }
word1 = "word"
word2 = "drow"
assert { reverse_words("#{word1} #{word2}") == "#{word1.reverse} #{word2.reverse}" }

# 5. Reflection 

# Not much to say about this exercise, we flew through it, wasn't much of a challenge.  It was a nice warm up to before moving
# on to something more difficult.

File: week_8_and_9/2_JavaScript/translate_to_js/my_solution.js
// U3.W8-9: Challenge you're converting: Reverse Words from Week 8/9


// I worked on this challenge [by myself].

// 2. Pseudocode

// write a function that accepts a string argument and returns that same string, but reversed.


// 3. Initial Solution

function reverseWords(string) {
	var newstring = string.split(" ");
	var revstring = []
	for (i=0; i < newstring.length; i++) {
		revstring[i] = newstring[i].split("").reverse().join("");
	}
	return revstring.join(" ");
}


// 4. Refactored Solution



// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

assert ( 
	reverseWords("Ich bin ein Berliner") === "hcI nib nie renilreB",
	"The sentance should be reversed!",
	"1. " 
	)

assert ( 
	reverseWords("") === "", 
	"The sentance should be empty!", 
	"2. " 
	)
assert (
	reverseWords("word") === "drow",
	"the word should have been reversed!",
	"3. "
	)


// 5. Reflection 

// Ruby is SO much easier to work with than JS.  This took me almost twice the amount of time to do.  That being said, I
// probably would have solved this much faster if I had more JS practice, so I blame me.  I don't think I can refactor this
// one much further.